### PR TITLE
Add FontFamily and FontCollection tests

### PR DESCRIPTION
**Pull request by hughbe**

### CODE REVIEW

**Comment by hughbe** [comment]
This commits 2 font files. The first is a simple OTF font, and the second is the same font - this time a TTF. I've also modified that file using a font editor to add extra features to test, such as custom language names embedded within the TTF file.

These files are licensed under creative commons (https://fontlibrary.org/en/font/code-new-roman).

Also, they add 600kb to the size of the repo. I think moving these to a dedicated repo may be a good idea at some point, either in this PR or at a later point. @stephentoub mentioned this

**Comment by stephentoub** [comment]
> I think moving the resources to a dedicated repo may be a good idea at some point, either in this PR or at a later point.

The problem is that once they're in the repo, they're there forever.  Even if you delete them, you're just adding a commit that removes them, but they're still there in the history.  So even if you delete them, everyone cloning the repo from here on out will still be pulling down the full size.  That's why this approach of adding binary files to the repo is bad.

**Comment by mellinoe** [comment]
I'll go ahead and create a nuget package over in https://github.com/dotnet/corefx-testdata so we can avoid adding any more images to the repo.

**Comment by mellinoe** [comment]
https://github.com/dotnet/corefx-testdata/pull/14

**Comment by mellinoe** [comment]
I merged in the change to use the TestData nuget package, so the new test assets will need to be submitted over in https://github.com/dotnet/corefx-testdata first.

**Comment by hughbe** [comment]
Yeah, do merge your Pr first - it'll also save me a lot of trouble if I have time to add tests today :)

Does the library instantly update or is there a long turnaround between adding a resource and being able to consume it in corefx

**Comment by mellinoe** [comment]
> Does the library instantly update or is there a long turnaround between adding a resource and being able to consume it in corefx

We don't have any automation for that repository, so every time we add stuff, I (or someone else on the team) will need to manually build the nuget package and publish it to myget. After that we can use it immediately in corefx.

**Comment by hughbe** [comment]
> /opt/corefx/Tools/dotnetcli/sdk/2.0.0-preview1-005977/NuGet.targets(97,5): error : Unable to resolve 'System.Drawing.Common.TestData (>= 1.0.1)' for '.NETStandard,Version=v2.0'. [/opt/corefx/external/test-runtime/XUnit.Runtime.depproj] [/opt/corefx/external/test-runtime/XUnit.Runtime.depproj]

Let's try again...

**Comment by hughbe** [comment]
@dotnet-bot test this please

**Comment by mellinoe** [comment]
I tried uploading it again, using a different method. It should work this time. I think I didn't follow the instructions correctly the first time (I'm at home and was trying to do it from memory).

**Comment by hughbe** [comment]
@dotnet-bot test this please

File: src/System.Drawing.Common/tests/FontFamilyTests.cs [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        [InlineData(-1, "Code New Roman")]
+        [InlineData(0, "Code New Roman")]
+        [InlineData(int.MaxValue, "Code New Roman")]
+        // This font has been modified to change the name to "Bonjour" if the language is French.
+        [InlineData(FrenchLCID, "Bonjour")]
+        public void GetName_LanguageCode_ReturnsExpected(int languageCode, string expectedName)
+        {
+            using (var fontCollection = new PrivateFontCollection())
+            {
+                fontCollection.AddFontFile(Helpers. GetTestFontPath("CodeNewRoman.ttf"));
```
Comment (mellinoe) [inline src/System.Drawing.Common/tests/FontFamilyTests.cs]: nit: space

File: src/System.Drawing.Common/tests/FontFamilyTests.cs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+                Assert.Equal(expectedName, fontFamily.Name);
+            }
+        }
+
+        [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotWindowsNanoServer))]
+        public void Ctor_Name_FontCollection()
+        {
+            using (var fontCollection = new PrivateFontCollection())
+            {
+                fontCollection.AddFontFile(Helpers. GetTestFontPath("CodeNewRoman.otf"));
```
Comment (mellinoe) [inline src/System.Drawing.Common/tests/FontFamilyTests.cs]: nit: space

File: src/System.Drawing.Common/src/System/Drawing/Design/CategoryNameCollection.cs [hunk 3]
```diff
@@ -3,68 +3,23 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections;
+using System.Security.Permissions;
 
 namespace System.Drawing.Design
 {
-    /// <summary>
```
Comment (hughbe) [inline src/System.Drawing.Common/src/System/Drawing/Design/CategoryNameCollection.cs]: It's my understanding we don't want doc comments like this in corefx for many reasons, the main one being that they're not used in the final product and we have the burden of maintaining 2 sets of docs

That said it wasn't deliberate that I removed these doc comments here - what do you think?

File: src/System.Drawing.Common/src/System/Drawing/FontFamily.cs [hunk 4]
```diff
@@ -236,14 +160,8 @@ private void Dispose(bool disposing)
                     Debug.Assert(status == SafeNativeMethods.Gdip.Ok, "GDI+ returned an error status: " + status.ToString(CultureInfo.InvariantCulture));
 #endif
                 }
-                catch (Exception ex)
+                catch (Exception ex) when (!ClientUtils.IsCriticalException(ex))
```
Comment (hughbe) [inline src/System.Drawing.Common/src/System/Drawing/FontFamily.cs]: We've changed this pattern before after discussing with @stephentoub in a previous PR. I think we reached the conclusion it has no functional difference as an exception is thrown if it's critical, but caught If it is not

File: src/System.Drawing.Common/src/System/Drawing/Design/CategoryNameCollection.cs (hunk 3 above)
Comment (mellinoe) [inline src/System.Drawing.Common/src/System/Drawing/Design/CategoryNameCollection.cs]: I'd rather keep them. We don't actually maintain any other doc comments, and ideally we will start producing and shipping the documentation from these source files. We do ship the comments from some libraries, like System.Numerics.Vectors. Other libraries don't even ship XML docs even though the source code has them 😞 .

Comment (hughbe) [inline src/System.Drawing.Common/src/System/Drawing/Design/CategoryNameCollection.cs]: I reverted the change to the doc comments.
> We don't actually maintain any other doc comments

What about https://docs.microsoft.com/en-us/dotnet/api/system.drawing.fontfamily?view=netframework-4.7

These docs seem to be a lot better and more comprehensive that the docs comments in the source

**Comment by hughbe** [comment]
Test UWP CoreCLR x64 Debug Build

**Comment by hughbe** [comment]
I've addressed PR feedback, mainly bringing back the doc comments

File: src/System.Drawing.Common/src/System/Drawing/FontFamily.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/FontFamily.cs]: For all intents and purposes it's the same.  There's some sublety around stack unwinding, but for our purposes the change is fine.

Presumably the Debug.Fail was removed because this is actually catching exceptions in our tests?

File: src/System.Drawing.Common/src/System/Drawing/Text/InstalledFontCollection.cs [hunk 5]
```diff
@@ -12,15 +12,10 @@ public sealed class InstalledFontCollection : FontCollection
         /// <summary>
         /// Initializes a new instance of the <see cref='System.Drawing.Text.InstalledFontCollection'/> class.
         /// </summary>
-        public InstalledFontCollection()
+        public InstalledFontCollection() : base()
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Text/InstalledFontCollection.cs]: Why did you add `: base()`?

File: src/System.Drawing.Common/src/System/Drawing/Text/PrivateFontCollection.cs [hunk 6]
```diff
@@ -16,45 +16,35 @@ public sealed class PrivateFontCollection : FontCollection
         /// <summary>
         /// Initializes a new instance of the <see cref='System.Drawing.Text.PrivateFontCollection'/> class.
         /// </summary>
-        public PrivateFontCollection()
+        public PrivateFontCollection() : base()
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Text/PrivateFontCollection.cs]: Why did you add `: base()`?

File: src/System.Drawing.Common/src/System/Drawing/FontFamily.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Drawing.Common/src/System/Drawing/FontFamily.cs]: Aye

File: src/System.Drawing.Common/src/System/Drawing/Text/InstalledFontCollection.cs (hunk 5 above)
Comment (hughbe) [inline src/System.Drawing.Common/src/System/Drawing/Text/InstalledFontCollection.cs]: It removes the redundant initialization of _nativeFontCollection to IntPtr.Zero

> internal FontCollection() => _nativeFontCollection = IntPtr.Zero;

File: src/System.Drawing.Common/src/System/Drawing/Design/CategoryNameCollection.cs (hunk 3 above)
Comment (mellinoe) [inline src/System.Drawing.Common/src/System/Drawing/Design/CategoryNameCollection.cs]: There are other doc comments, but they aren't maintained by the devs on our team, so they are a lot more difficult to change. They also don't show up when you are working on the library itself, which is the main benefit to these comments.

**Comment by mellinoe** [comment]
Thanks for the contribution, @hughbe !
